### PR TITLE
LEX-1836 Create unauthenticated Client and use it for Capture.

### PR DESCRIPTION
**Pull request by ynx**

### CODE REVIEW

File: src/Lift.php [hunk 1]
```diff
@@ -15,25 +15,20 @@
 use GuzzleHttp\Psr7\Request;
 use Psr\Http\Message\RequestInterface;
 
-class Lift extends Client
+class Lift
```
Comment (ynx) [inline src/Lift.php:18]: Lift class is no longer a Client itself. Instead, it employs two Clients.

File: src/Lift.php [hunk 2]
```diff
@@ -60,38 +55,43 @@ public function __construct(
...
-        // Set our default HandlerStack if nothing is provided
-        if (!isset($config['handler'])) {
-            $config['handler'] = HandlerStack::create();
-        }
-
-        // Add our Account and Site identifiers.
-        $config['handler']->push($this->addLiftAccountAndSiteId());
+        // Both stacks are cloned so they are both completely internal now.
+        $unauthenticatedHandlerStack = clone $handlerStack;
+        $authenticatedHandlerStack = clone $handlerStack;
```
Comment (ynx) [inline src/Lift.php:65]: I could get away with cloning only once, but I wanted to keep consistent behaviour for both handler stacks.

File: src/Lift.php [hunk 3]
```diff
@@ -118,16 +119,15 @@ public function addLiftAccountAndSiteId()
...
      */
     public function ping()
     {
-        // Now make the request.
         $request = new Request('GET', '/ping');
+        $response = $this->authenticatedClient->send($request);
+        $body = (string) $response->getBody();
 
-        return $this->getResponseJson($request);
+        return json_decode($body, true);
```
Comment (ynx) [inline src/Lift.php:130]: There is a bit of duplicated code here as opposed to ManagerBase class's code, but I think it is too much to create a ping manager or move any code to a trait.

File: src/Lift.php [hunk 4]
```diff
@@ -137,7 +137,7 @@ public function ping()
      */
     public function getSlotManager()
     {
-        return new SlotManager($this);
+        return new SlotManager($this->authenticatedClient);
```
Comment (ynx) [inline src/Lift.php:140]: Notice this is now ```authenticatedClient```.

File: src/Lift.php [hunk 5]
```diff
@@ -177,7 +177,7 @@ public function getRuleManager()
      */
     public function getCaptureManager()
     {
-        return new CaptureManager($this);
+        return new CaptureManager($this->unauthenticatedClient);
```
Comment (ynx) [inline src/Lift.php:180]: Notice this is now ```unauthenticatedClient```.

File: src/Manager/DecideManager.php [hunk 6]
```diff
@@ -22,9 +22,9 @@ class DecideManager extends ManagerBase
     public function decide(Decide $decide)
     {
         $body = $decide->json();
-        $url = '/capture';
+        $url = '/decide';
```
Comment (ynx) [inline src/Manager/DecideManager.php:25]: I think was a typo. Fixed here.

Comment (eshtadc) [reply src/Manager/DecideManager.php:25]: That's a pretty big typo (or copy/pasta).  Is there a test to catch this?

Comment (ynx) [reply src/Manager/DecideManager.php:25]: I say no. I am about to add some tests for this ticket so I will cover this "/decide" as well.

Comment (ynx) [reply src/Manager/DecideManager.php:25]: Tested.

File: src/Manager/RuleManager.php [hunk 7]
```diff
@@ -34,17 +47,11 @@ class RuleManager extends ManagerBase
     public function query($options = [])
     {
         $url = '/rules';
-        $url .= isset($options['visible_on_page']) ? "&visible_on_page={$options['visible_on_page']}" : '';
-        $url .= isset($options['prefetch']) ? "&prefetch={$options['prefetch']}" : true;
-        $url .= isset($options['sort']) ? "&sort={$options['sort']}" : 'asc';
-        $url .= isset($options['start']) ? "&start={$options['start']}" : 0;
-        $url .= isset($options['rows']) ? "&rows={$options['rows']}" : 10;
-        $url .= isset($options['sort_field']) ? "&sort_field={$options['sort_field']}" : 'updated';
-        $url .= isset($options['status']) ? "&status={$options['status']}" : 'published';
```
Comment (ynx) [inline src/Manager/RuleManager.php:43]: 1) This code is mechanically not working on many levels. It is fixed by use of a new function ```getQueryString()```.

2) Notice the original approach has hardcoded default options. I think it is a bad idea to have SDK specific default options, so I have removed those defaults and let the queries fallback to Decision API defaults. If we want to define query defaults, we should do it at Decision API.

File: test/CaptureTest.php [hunk 8]
```diff
@@ -128,6 +128,32 @@ private function setTestCapturesResponseData()
...
+
+        // Check if the client has already have expected handlers.
+        // To check, to insert a dummy function after the expected handler, and
+        // hope it finds the expected handler (or throw Exception when don't).
+        $handler = $manager->getClient()->getConfig('handler');
+        $testFunction = function () {};
+        $handler->after('acquia_lift_account_and_site_ids', $testFunction);
+        // Throws Exception because this handler is unauthenticated.
+        $handler->after('acquia_lift_hmac_auth', $testFunction);
+    }
```
Comment (ynx) [inline test/CaptureTest.php:155]: Unit testing to assert CaptureManager's handler is not authenticated.

File: test/CaptureTest.php [hunk 9]
```diff
@@ -137,11 +163,19 @@ public function testCaptureAdd()
...
 
-        // Get Rule Manager
+        // Get Capture Manager
         $manager = $client->getCaptureManager();
         $response = $manager->add($this->capturePayload);
+        $request = $this->mockHandler->getLastRequest();
+
+        // Check for request configuration
+        $this->assertEquals($request->getMethod(), 'POST');
+        $this->assertEquals((string) $request->getUri(), '/capture?account_id=TESTACCOUNTID&site_id=TESTSITEID');
```
Comment (ynx) [inline test/CaptureTest.php:173]: Tests the method and URL, as requested in last round of comments.

File: test/GoalTest.php [hunk 10]
```diff
@@ -7,6 +7,27 @@
...
+        $manager = $client->getRuleManager();
+
+        // Check if the client has already have expected handlers.
+        // To check, to insert a dummy function after the expected handler, and
+        // hope it finds the expected handler without throwing an Exception.
+        $handler = $manager->getClient()->getConfig('handler');
+        $testFunction = function () {};
+        $handler->after('acquia_lift_account_and_site_ids', $testFunction);
+        // Does not throw Exception because this handler is authenticated.
+        $handler->after('acquia_lift_hmac_auth', $testFunction);
```
Comment (ynx) [inline test/GoalTest.php:28]: Unit testing to assert GoalManager's handler is authenticated.

File: test/RuleTest.php [hunk 11]
```diff
@@ -277,9 +316,29 @@ public function testRuleQuery()
...
+            'sort_field' => 'updated',
+            'status' => 'unpublished',
+            'unrelated_option_name' => 'unrelated_option_value',
+        ];
+        $responses = $manager->query($option);
+        $request = $this->mockHandler->getLastRequest();
+
+        // Check for request configuration
+        $this->assertEquals($request->getMethod(), 'GET');
+        $this->assertEquals((string) $request->getUri(), '/rules?visible_on_page=my_page&prefetch=true&sort=desc&start=4&rows=20&sort_field=updated&status=unpublished&account_id=TESTACCOUNTID&site_id=TESTSITEID');
```
Comment (ynx) [inline test/RuleTest.php:336]: Testing query options fixes.

File: test/TestBase.php [hunk 12]
```diff
@@ -44,15 +52,28 @@ protected function setUp()
...
+     * Get a dummy auth middleware.
+     */
+    protected function getDummyAuthMiddleware()
+    {
+        return function (callable $handler) {
+            return function (RequestInterface $request, array $options) use ($handler) {
+                return $handler($request, $options);
+            };
+        };
+    }
```
Comment (ynx) [inline test/TestBase.php:80]: We set in a dummy auth middleware, which does nothing. It replaces the real auth middleware, so we don't have to further complicate other tests.

File: src/Manager/ManagerBase.php [hunk 13]
```diff
@@ -40,4 +54,28 @@ protected function getResponseJson(RequestInterface $request)
...
+     *
+     * @param $options The options
+     * @return string  The query string
+     */
+    protected function getQueryString($options) {
+      $queries = [];
+      foreach ($this->queryParameters as $queryName => $queryDefaultValue) {
+        // Use user value if possible.
+        if (isset($options[$queryName])) {
+          $queries[] = $queryName . '=' . $options[$queryName];
```
Comment (eshtadc) [inline src/Manager/ManagerBase.php]: Are these already encoded?  Do you have to worry about that at all?

File: src/Manager/ManagerBase.php [hunk 14]
```diff
@@ -40,4 +54,28 @@ protected function getResponseJson(RequestInterface $request)
...
+          continue;
+        }
+        // Use default value if possible.
+        if (!isset($options[$queryName]) && is_string($queryDefaultValue)) {
```
Comment (eshtadc) [inline src/Manager/ManagerBase.php]: Doesn't seem like you need the `!isset($options[$queryName])` check as it is checked above and followed by a `continue`.

File: src/Manager/RuleManager.php (hunk 7 above)
Comment (eshtadc) [reply src/Manager/RuleManager.php:43]: I agree - but maybe we should call this out in the examples?  The biggest one is "prefetch" as it defaults to `false` in the API.

Comment (eshtadc) [reply src/Manager/RuleManager.php:43]: Actually - nvm... just checked the examples and they already pass the options in.... so yes - I totally agree with you :)

File: test/TestBase.php (hunk 12 above)
Comment (eshtadc) [reply test/TestBase.php:80]: That seems worthy of putting in the comments there.

Comment (ynx) [reply test/TestBase.php:80]: Done @ 1a8864b0dc329932add2939a384a840b648b1f58.

File: src/Manager/ManagerBase.php (hunk 14 above)
Comment (ynx) [reply src/Manager/ManagerBase.php]: You are right. Done @ 0d722f29388f753ddb955c9ecdabfe19d737fee5.

File: src/Manager/ManagerBase.php (hunk 13 above)
Comment (ynx) [reply src/Manager/ManagerBase.php]: Nice catch. I do need to worry about it. Done @ ea2057291ad686cbe7b3c4cc0fab6e94cd0d7e47.

**Pull request merged**